### ISSUE TITLE
Please add aiplatform.log_artifact() or aiplatform.ExperimentRun.log_artifact()

### ISSUE THREAD

**Issue by jasonbrancazio** [googleapis/python-aiplatform#1502]
Is there some way to use the new classes and methods introduced in aiplatform==1.15.0 to accomplish something like this?

aiplatform.ExperimentRun.log_artifact(Artifact)

Background:

[Sometimes](https://engineering.sada.com/machine-learning-orchestration-on-vertex-ai-3fb8cb9b787a) you don't need a full-fledged Vertex AI pipeline to accomplish a training task. A Custom Job or Vertex AI Training Pipeline suffices. In these cases, you might still wish to log experiment data. A simple and common task might be to train a model, report the model training parameters and metrics to Vertex AI Experiments, save the model to cloud storage, and associate the experiment run with an artifact representing the saved model.

I have been able to accomplish the metric and parameter reporting within a CustomJob using a [service account](https://github.com/googleapis/python-aiplatform/issues/902) with the appropriate scope and aiplatform.log_params and aiplatform.log_metrics for some time. Even after the amazing improvements in aiplatform==1.15.0, however, I don't see a similar way to log an artifact. aiplatform.log() only permits logging a PipelineJob.

Am I mistaken? 

Might some workaround exist that would involve manually creating entries in the metadata store?

Regardless, aiplatform.log_artifact() or aiplatform.ExperimentRun.log_artifact() or aiplatform.log(artifact: Artifact) would be a welcome improvement to the client library.

**Comment by jaycee-li** [comment]
Yes, we currently don't support logging an artifact directly to the ExperimentRun. But you can log artifacts through Executions. Here is the steps (assuming you already have an artifact for dataset `train_dataset` and an artifact for model `my_model`)
```
aiplatform.start_run('my_run')
with aipaltform.start_execution('train') as exc:
    exc.assign_input_artifacts([train_dataset])
    exc.assign_output_artifacts([my_model])

## train-dataset --> train --> my_model
```
By this way, all the artifacts are associated with the ExperimentRun, and you can track the lineage through the execution (`my_model` was trained on `train_dataset`).

**Comment by jasonbrancazio** [comment]
Thanks for the feedback. I experimented some more with the list() methods of the _Context, Execution, and Artifact classes to understand the effects of your recommended code in the metadata store. 

I now see that Experiments and ExperimentRuns are Contexts, and that aiplatform.log_params and log_metrics create metadata of Contexts. Artifacts, however, need to be associated with Executions. 

You can probably close this feature request because it likely won't make sense to implement.

**Comment by jasonbrancazio** [comment]
For a new user, `log_artifact` still makes a lot of sense. Once a person uses `log_params` and `log_metrics`, and then visits the new Vertex AI cloud console for Experiments, they will see that Artifacts can in fact be associated with ExperimentRuns. Then they evaluate the aiplatform client library to learn how to do so, and it's not as easy. 

For the future reader of this github issue, in addition to @jaycee-li's code above, you should check out this post for more examples:
https://cloud.google.com/blog/topics/developers-practitioners/track-compare-manage-experiments-vertex-ai-experiments

**Comment by yinghsienwu** [comment]
Thanks for the feedback.
